### PR TITLE
style: Support "auto" keyword on min-{width, height} properties

### PR DESCRIPTION
**Pull request by stshine**
Based on the work of #14332 .

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15969)

### CODE REVIEW

**Comment by stshine** [comment]
Hmm.
@bors-servo try

**Comment by stshine** [comment]
The problem obviously exists [here](https://github.com/servo/servo/pull/15969/files#diff-f488bae1b885cd00ad5abe2ebb8631a0L354), but it reveals that maybe we do not handle `calc()` correctly when the container size is infinite. Let me take a more careful look before landing this.

**Comment by stshine** [comment]
The problem I found after sbwtw's patch is the `specified*()` functions in `model.rs` accept `Au` as container size, and when they are used to calculate the intrinsic size of an element an `Au(0)` is passed, which is wrong. This need some work and I haven't figure out a clean fix for it, but I think I can finish this first.

**Comment by jdm** [comment]
```
error[E0425]: cannot find function `adjust_auto_minsize` in module `properties`
   --> /home/servo/buildbot/slave/linux-dev/build/components/style/style_adjuster.rs:263:9
    |
263 |         properties::adjust_auto_minsize(self.style, is_item);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `properties`
```

**Comment by jdm** [comment]
```
  ▶ FAIL [expected PASS] /css-values-3_dev/html/calc-height-table-1.htm
  └   → /css-values-3_dev/html/calc-height-table-1.htm 4a252b2743413b7bc402d2f276f964841ac579f0
/css-values-3_dev/html/reference/calc-height-table-1-ref.htm 1081f8dadc3529ed5cef96ecdf02d7305e895d20
Testing 4a252b2743413b7bc402d2f276f964841ac579f0 == 1081f8dadc3529ed5cef96ecdf02d7305e895d20

  ▶ FAIL [expected PASS] /css-values-3_dev/html/calc-min-height-block-1.htm
  └   → /css-values-3_dev/html/calc-min-height-block-1.htm 18f3b6eb8cf23dd9207d4ed25a37bf45d5390284
/css-values-3_dev/html/reference/calc-height-block-1-ref.htm e38f694bf79dd9135a9b5742c7f681e5abea82a7
Testing 18f3b6eb8cf23dd9207d4ed25a37bf45d5390284 == e38f694bf79dd9135a9b5742c7f681e5abea82a7
```

**Comment by stshine** [comment]
Woooo newly passed tests. @emilio this patch is really for review, but I'm not sure if you can accept the ugly status of it :)

File: components/style/properties/properties.mako.rs [hunk 1]
```diff
@@ -2571,6 +2571,9 @@ pub fn apply_declarations<'a, F, I>(device: &Device,
         }
     % endif
 
+    % if product == "servo":
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: Remove this empty block?

File: components/style/properties/properties.mako.rs [hunk 2]
```diff
@@ -2588,6 +2591,20 @@ pub fn apply_declarations<'a, F, I>(device: &Device,
...
 
+/// See StyleAdjuster::adjust_for_auto_minsize.
+% if product == "servo":
+pub fn adjust_auto_minsize(style: &mut StyleBuilder, is_item: bool) {
+    % for direction in ["width", "height"]:
+        // Like calling to_computed_value, which wouldn't type check.
+        if style.get_position().clone_min_${direction}() == computed::LengthOrPercentageOrAuto::Auto
+            && !is_item {
+                style.mutate_position().set_min_${direction}(
+                    computed::LengthOrPercentageOrAuto::Length(Au(0))
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: I think a spec quote/link in `StyleAdjuster::adjust_for_auto_minsize` would be neat.

File: tests/wpt/mozilla/tests/css/flex-auto-min-size-fallback.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
+<html><head><title>non-flexbox | computed style | min-width: auto</title>
+<link href="http://www.w3.org/TR/css-flexbox-1/#min-size-auto" rel="help">
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/flex-auto-min-size-fallback.html]: Please don't use `TR` links for CSS specs, use https://drafts.csswg.org/css-flexbox/.

File: components/style/properties/properties.mako.rs (hunk 2 above)
Comment (emilio) [inline components/style/properties/properties.mako.rs]: This is not completely correct AFAICT. This also depends on the `overflow` property.

The following test-case:

```html
<!doctype html>
<style>
.item {
  min-width: auto;
  min-height: auto;
  overflow: visible;
}
</style>
<div id="log"></div>
<div style="display: flex">
  <div class="item"></div>
</div>
<script>
var item = document.querySelector('.item');
alert(getComputedStyle(item).minHeight)
alert(getComputedStyle(item).minWidth)
</script>
```

Alerts `auto`, `auto` in chrome, but `0px`, `auto` in Firefox.

The relevant bit in Firefox is here (FF handles this in `ComputedStyleDeclaration`): http://searchfox.org/mozilla-central/rev/abe68d5dad139e376d5521ca1d4b7892e1e7f1ba/layout/style/nsComputedDOMStyle.cpp#5079

I'll look for the relevant bit in Blink.

Comment (emilio) [inline components/style/properties/properties.mako.rs]: Sorry, the comment above doesn't represent well the dependence on overflow, though it's obvious reading Gecko's source.

The following test-case:

```html
<!doctype html>
<style>
.item {
  min-width: auto;
  min-height: auto;
  overflow: hidden;
}
</style>
<div id="log"></div>
<div style="display: flex">
  <div class="item"></div>
</div>
<script>
var item = document.querySelector('.item');
alert(getComputedStyle(item).minHeight)
alert(getComputedStyle(item).minWidth)
</script>
```

Alerts `0px`, `0px` on FF, but `auto`, `auto` on Blink. I'm looking for the specs to see who's the one right here.

Comment (emilio) [inline components/style/properties/properties.mako.rs]: Seems like Gecko is right here per https://drafts.csswg.org/css-flexbox/#valdef-min-width-auto:

> On a flex item whose overflow is visible in the main axis, when specified on the flex item’s main-axis min-size property, specifies an automatic minimum size.

I think it's reasonable to do this in CSSStyleDeclaration, (or `process_resolved_style_query`)

Comment (stshine) [inline components/style/properties/properties.mako.rs]: My last try failed to find where the spec asked to return `0px` in normal element ...

Comment (emilio) [inline components/style/properties/properties.mako.rs]: Here's the blink code, also only lying on `getComputedStyle`:

https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/css/ComputedStyleCSSValueMapping.cpp?l=2705&rcl=4e51594d2e742b0317f507f21d483d75dd25e1e6

Comment (stshine) [inline components/style/properties/properties.mako.rs]: Guh, `ThreadSafeLayoutNode` does not allow access to parent node. Do you have an idea how to know if the parent is a flex flow?

Comment (emilio) [inline components/style/properties/properties.mako.rs]: Presumably propagating a bit down the flow tree? We could also use a custom CSS property I guess, though ideally we'd just use the flow tree.

**Comment by emilio** [comment]
Closing due to inactivity, feel free to reopen / submit a new PR if you intend to keep working on this @stshine, thanks a lot!

**Comment by stshine** [comment]
Sorry I'm having a hard time figuring out how caching works in WebRender, will back to this later.

**Pull request closed without merging**